### DISCUSSION TITLE
Webquery or Ping for web service - how to?

### DISCUSSION THREAD

**Discussion by tanker11** [arendst/Tasmota#20389] [Q&A]
Hello,

I am trying to put together a script to check whether an IP address answers a ping or a webquery.
I could not find a working solution to use the response in variable and test if the response is
RESULT = {"WebQuery":"Done"} or 
RESULT = {"WebQuery":"Connect failed"}

I cannot find how to test if there is no any json answer expected only if the availability of the service matters.
Definitions in user_config:

#define USE_SCRIPT
#define USE_WEBSEND_RESPONSE
#define USE_SCRIPT_STATUS
#define USE_PING
#define USE_RULES

**Comment by barbudor** [comment]
#define USE_SCRIPT
...
#define USE_RULES

That's not going to work
Rules and Scripting-Language are exclusive

**Comment by tanker11** [comment]
Sorry, I was not correct. The relevan def section is like as below, so rules are not included:

#define USE_SCRIPT

#ifdef USE_RULES
#undef USE_RULES
#endif 

#define USE_WEBSEND_RESPONSE
#define USE_SCRIPT_STATUS
#define USE_PING

And still, I have no usable response in a variable...

**Reply by barbudor** [reply]
I have no knowledge on Scripting language, but as a general advice posting the script that you have done has higher chance to attract comments
And to make code-paste readable, use the triple back-tick :
![image](https://github.com/arendst/Tasmota/assets/2996042/314e504b-64dd-47f9-9bf4-68a970d0d502)
=>
```
like this
```

**Reply by tanker11** [reply]
You are right, better to post what I am trying to achieve with the above depicted situation:

```
>D 48

pos=0
result=""
result2=""

>S
if upsecs%4==0
then
print Counter: %0pos%
=>WebQuery http://192.168.0.1/ GET
;=>WebSend [192.168.0.1:80]
;=>PING 192.168.0.1
result=WebQuery
print EVENT: %result%
pos=pos+1
endif

>E
print event called!
result2=WebQuery
print EVENT: %result2%


```

The result in the console is the below:

22:33:59.414 SCR: performs "WebQuery http://192.168.0.1/ GET"
22:33:59.451 RSL: RESULT = {"WebQuery":"Done"}
22:33:59.453 EVENT: 

It is visible that that variable is not filled with the result and not even the >E section called.

**Reply by sfromis** [reply]
I'm not using scripting, but your output looks like you got no response to that `WebQuery` command, which might imply that you're not using a build with `USE_WEBSEND_RESPONSE`. When having that in a build (not a scripting one), I'm seeing the actual response instead of "Done".

**Comment by tanker11** [comment]
Actually, I am using a build with USE_WEBSEND_RESPONSE.
And if I use the console with ```WebQuery http://192.168.0.1/ GET``` it has the correct output:

23:27:01.705 CMD: WebQuery http://192.168.0.1/ GET
23:27:01.727 RSL: RESULT = {"WebQuery":"Done"}
23:27:01.729 event called!
23:27:01.732 EVENT: Done

**Reply by sfromis** [reply]
The intended result with `USE_WEBSEND_RESPONSE` is that you'd be getting the actual response, instead of merely "Done". Like this from a minimal test I just did: `{"WebQuery":"test"}`
That is what I see as "the correct output".

**Reply by tanker11** [reply]
Well, in my case, the purpose of webquery is only to see there is a webpage on the given IP address, so I do not expect anything in the answer. As you can see above, if I use webquery command in the console, I get the response (and the >E section is called and extracts the expected "Done" from the JSON answer. My problem is that if I trigger the command from the script, nor the >E section is called nor the expected result is filled into my variable. do you still think USE_WEBSEND_RESPONSE is not the part of my build? How could I check/prove it to find out if this has something to do with the unexpected result?

**Reply by sfromis** [reply]
As the console message you post about getting does not include the response, there is nothing to indicate that `USE_WEBSEND_RESPONSE` is included. And IDK if that would make any difference to using the scripter feature, which I'm not using.

To be clear, do not read my words above as a diagnosis about `USE_WEBSEND_RESPONSE` certainly not being included, just that my experience is that when doing a simple http request with that feature enabled, and getting a short response text, I'm seeing that reported and not just "Done". But this may depend on what type of response is returned from the server, like if it is empty or not.

When it comes to your observation about it only failing when you invoke the command from a script, I'm reminded of a similar situation with rules, which only can pick up the response from rule-issued commands if they are "wrapped" in a Backlog command. Maybe try that.

**Reply by tanker11** [reply]
Thank you for the clarification and dealing with the issue.
Finally I could make a breaktrhough, the compiled items were OK, the only change I needed to make is that I used different command call, namely:
```+>WebQuery http://192.168.0.1/ GET```
instead of the above quoted one:
```=>WebQuery http://192.168.0.1/ GET```

Then, >E section is called properly, and the result is filled into the variable.
Frankly, based on the scripting docs I cannot make a good reasoning why (docs: https://tasmota.github.io/docs/Scripting-Language/#commands )

So this is a typical situation that it works but I don't know why... never mind, I could proceed with my code and it works now. Thanks again!

**Reply by pkkrusty** [reply]
You can also use rules. See below:

```
Rule1 
ON Time#Initialized DO backlog ruletimer2 10; var2 1 ENDON 
on rules#timer==2 do backlog add2 1; ruletimer2 8; event t1=%var2% endon 
on var2#state>=12 do var2 1 endon 
on event#t1==1 do backlog ping4 192.168.3.70; var3 MoringaRouter endon 
on event#t1==2 do backlog ping4 192.168.3.9; var3 DortoirRouter endon 
on event#t1==3 do backlog ping4 192.168.3.22; var3 ClinicRouter endon 
on event#t1==4 do backlog ping4 192.168.3.14; var3 CouvoirRouter endon 
on event#t1==5 do backlog ping4 192.168.3.33; var3 MSRouter endon 
on event#t1==6 do backlog ping4 192.168.3.53; var3 GuardRouter endon 
on event#t1==7 do backlog ping4 192.168.3.29; var3 JohannesRouter endon 
on event#t1==8 do backlog ping4 192.168.3.64; var3 SchoolRouter endon 
on event#t1==9 do backlog ping4 192.168.3.71; var3 AtelierRouter endon 
on event#t1==10 do backlog ping4 192.168.3.52; var3 DogoRouter endon 
on event#t1==11 do backlog ping4 192.168.3.11; var3 BouchRouter endon 
on ping#?#Success do WebQuery http://192.168.3.48:8086/write?db=homeautomation&rp=four_weeks POST ping,device=%var3% value=%value% ENDON
```
Writing ping results to influxdb.

**Comment by tanker11** [comment]
Thanks for that, very useful!

pkkrusty ***@***.***> ezt írta (időpont: 2024. jan. 5., P,
22:27):
